include/ipc/ipc_service.h

```diff
@@ -157,21 +222,154 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  @param[in] data Pointer to the TX buffer.
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *		 backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *  @retval -EALREADY when the buffer was already dropped.
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_drop_tx_buffer(struct ipc_ept *ept, const void *data);
+
+/** @brief Send data in a TX buffer reserved by @ref ipc_service_get_tx_buffer
+ *         using the given IPC endpoint.
+ *
+ *  This is equivalent to @ref ipc_service_send but in this case the TX buffer
+ *  has been obtained by using @ref ipc_service_get_tx_buffer.
+ *
+ *  The application has to take the responsibility for getting the TX buffer
+ *  using @ref ipc_service_get_tx_buffer and filling the TX buffer with the data.
+ *
+ *  After the @ref ipc_service_send_nocopy function is issued the TX buffer is
+ *  no more owned by the sending task and must not be touched anymore unless
+ *  the function fails and returns an error.
+ *
+ *  If this function returns an error, @ref ipc_service_drop_tx_buffer can be
+ *  used to drop the TX buffer.
+ *
+ *  @param[in] ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param[in] data Pointer to the buffer to send obtained by @ref
+ *		    ipc_service_get_tx_buffer.
+ *  @param[in] len Number of bytes to send.
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *		 backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -EBADMSG when the data is invalid (i.e. invalid data format,
+ *		     invalid length, ...)
+ *  @retval -EBUSY when the instance is busy.
+ *
+ *  @retval bytes number of bytes sent.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_send_nocopy(struct ipc_ept *ept, const void *data, size_t len);
+
+/** @brief Release the RX buffer for future reuse.
+ *
+ *  When supported by the backend, this function can be called after the
+ *  received message has been processed and the buffer can be marked as
+ *  reusable again.
+ *
+ *  @param[in] ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param[in] data Pointer to the RX buffer to release.
+ *
+ *  @retval -EIO when no backend is registered or release hook is missing from
+ *		 backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -EALREADY when the buffer data has been already released.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_release_rx_buffer(struct ipc_ept *ept, const void *data);
+
+/** @brief Holds the RX buffer for usage outside the receive callback.
```

Comment:
It would be easier to follow this API if `hole_rx_buffer()` was described before `release_rx_buffer()`.

I'm missing an information here if implementation is supposed to count number of `hold_rx_buffer()` calls and release the buffer only when matching number of `release_rx_buffer()` is called, or is it like a binary status.

Let me show an example:
```
void received_callback(...) {
  ipc_service_hold_rx_buffer(...);
  ...
}

void other_processing(void) {
  ipc_service_hold_rx_buffer(...);

  ipc_service_release_rx_buffer(...); // Is buffer released here?
  ipc_service_release_rx_buffer(...); // Or here?
}
```

Reply:
so
1) `hold` and `release` are acting on the `data` parameter:
```
ipc_service_hold_rx_buffer(A);
ipc_service_hold_rx_buffer(B);
...
ipc_service_release_rx_buffer(B); // This is releasing buffer B
ipc_service_release_rx_buffer(A); // This is releasing buffer A
```
2) You cannot release the same buffer multiple times, otherwise you get `-EALREADY`:
```
ipc_service_hold_rx_buffer(A);
ipc_service_hold_rx_buffer(A); # here you get -EALREADY
...
ipc_service_release_rx_buffer(A);
ipc_service_release_rx_buffer(A); # here you get -EALREADY
```

I will add a new error case for when you try to `release` a buffer that was not `hold`-ed.

Reply:
> I'm missing an information here if implementation is supposed to count number of hold_rx_buffer() calls and release the buffer only when matching number of release_rx_buffer() is called, or is it like a binary status.

I mentioned it earlier. I would add restriction that there is only one buffer being held and only by one owner. Otherwise this will become complex. Note that practically there will be always one context using the buffer.

Reply:
I meant calling `hold_rx_buffer()` multiple times for the same buffer, so restricting number of buffers would not answer my question.

But `-EALREADY` returned by `hold_rx_buffer()` clarifies it.